File: backend/mycrm/src/main/java/parrino/riccardo/mycrm/service/SaleService.java
package parrino.riccardo.mycrm.service;

import java.util.Date;
import java.util.List;
import java.util.Optional;

import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.stereotype.Service;

import parrino.riccardo.mycrm.dto.SaleDTO;
import parrino.riccardo.mycrm.model.Customer;
import parrino.riccardo.mycrm.model.Product;
import parrino.riccardo.mycrm.model.Sale;
import parrino.riccardo.mycrm.model.SaleId;
import parrino.riccardo.mycrm.model.User;
import parrino.riccardo.mycrm.repository.CustomerRepository;
import parrino.riccardo.mycrm.repository.ProductRepository;
import parrino.riccardo.mycrm.repository.SaleRepository;
import parrino.riccardo.mycrm.repository.UserRepository;

@Service
public class SaleService {

    private final UserRepository userRepository;
    private final SaleRepository saleRepository;
    private final CustomerRepository customerRepository;
    private final ProductRepository productRepository;
    private final UserService userService;

    public SaleService(
        UserRepository userRepository,
        SaleRepository saleRepository,
        CustomerRepository customerRepository,
        ProductRepository productRepository,
        UserService userService
    ) {
        this.userRepository = userRepository;
        this.saleRepository = saleRepository;
        this.customerRepository = customerRepository;
        this.productRepository = productRepository;
        this.userService = userService;
    }


    public Boolean createSale(SaleDTO saleDTO) {
        try {
            String username = SecurityContextHolder.getContext().getAuthentication().getName();
            Optional<User> user = userService.findUserByUsername(username);

            if (
                user.isPresent() &&
                user.get().getCustomers().indexOf(Customer.builder().customerId(saleDTO.getCustomerId()).build()) >= 0 &&
                user.get().getProduce().indexOf(Product.builder().productId(saleDTO.getProductId()).build()) >= 0
            ) {
                Sale sale = Sale.builder()
                    .saleId(
                        SaleId.builder()
                            .username(saleDTO.getUsername())
                            .customerId(saleDTO.getCustomerId())
                            .productId(saleDTO.getProductId())
                            .createdAt(new Date())
                        .build()
                    )
                    .user(user.get())
                    .customer(customerRepository.findById(saleDTO.getCustomerId()).get())
                    .product(productRepository.findById(saleDTO.getProductId()).get())
                    .progress(saleDTO.getProgress())
                    .activity(saleDTO.getActivity())
                    .amount(saleDTO.getAmount())
                    .lastUpdate(saleDTO.getLastUpdate())
                    .notes(saleDTO.getNotes())
                    .build();
                saleRepository.save(sale);
                return true;
            } else {
                return false;
            }
        } catch (Exception e) {
            System.out.println(e);
            return false;
        }
    }

    public List<Sale> readSale() {
        return saleRepository.findAll();
    }
    
    public Boolean updateSale(SaleDTO saleDTO) {
        if (saleRepository.existsById(SaleId.builder()
                            .saleId(saleDTO.getSaleId())
                            .username(saleDTO.getUsername())
                            .customerId(saleDTO.getCustomerId())
                            .productId(saleDTO.getProductId())
                            .createdAt(saleDTO.getCreatedAt())
                            .build())
        ) {
            Sale sale = Sale.builder()
                .saleId(
                    SaleId.builder()
                        .saleId(saleDTO.getSaleId())
                        .username(saleDTO.getUsername())
                        .customerId(saleDTO.getCustomerId())
                        .productId(saleDTO.getProductId())
                        .createdAt(new Date())
                    .build()
                )
                .user(userRepository.findById(saleDTO.getUsername()).get())
                .customer(customerRepository.findById(saleDTO.getCustomerId()).get())
                .product(productRepository.findById(saleDTO.getProductId()).get())
                .progress(saleDTO.getProgress())
                .activity(saleDTO.getActivity())
                .amount(saleDTO.getAmount())
                .lastUpdate(new Date())
                .notes(saleDTO.getNotes())
                .build();
            saleRepository.updateSale(
                sale.getCustomer().getCustomerId(),
                sale.getProduct().getProductId(),
                sale.getActivity(),
                sale.getProgress(),
                sale.getAmount(),
                sale.getLastUpdate(),
                sale.getNotes(),
                sale.getSaleId().getSaleId()
            );
            return true;
        }
        return false;
    }
    
    public Boolean deleteSale(SaleId saleId) {
        try {
            saleRepository.deleteById(saleId);
            return true;
        } catch(Exception e) {
            System.out.println(e);
            return false;
        }
    }

}

File: frontend/src/app/service/sale.service.ts
import { HttpClient, HttpParams } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { SaleDTO } from '../dto/sale.dto';
import { Sale } from '../model/sale.model';
import { SaleId } from '../model/saleId.model';

@Injectable({
  providedIn: 'root'
})
export class SaleService {

  readUrl : string = 'http://localhost:8080/sale/read'
  createUrl : string = 'http://localhost:8080/sale/create'
  updateUrl : string = 'http://localhost:8080/sale/update'
  deleteUrl : string = 'http://localhost:8080/sale/delete'

  constructor(private httpClient:HttpClient) {}

  readSales() : Observable<SaleDTO[]> {
    return this.httpClient.get<SaleDTO[]>(this.readUrl);
  }

  createSale(sale:Sale) : Observable<Boolean> {
    const saleDTO = {
      'saleId' : sale.saleId.saleId,
      'createdAt' : sale.saleId.createdAt,
      'customerId': sale.saleId.customerId,
      'productId': sale.saleId.productId,
      'progress': sale.progress,
      'activity': sale.activity,
      'amount': sale.amount,
      'lastUpdate': sale.lastUpdate,
      'notes': sale.notes
    };
    return this.httpClient.post<Boolean>(this.createUrl, saleDTO);
  }

  updateSale(sale:any) : Observable<Boolean> {
    console.log(sale);
    return this.httpClient.post<Boolean>(this.updateUrl, sale);
  }

  deleteSale(saleId:SaleId) : Observable<Boolean> {
    return this.httpClient.post<Boolean>(this.deleteUrl, saleId);
  }

}
